### PR TITLE
trendingUrl, jotLog validations, date range filter

### PR DESCRIPTION
**Pull request by shanmukhipriya-ponnada**
Tested in dev and working 
Screenshot 2023-02-07 at 12 16 04 PM

### CODE REVIEW

File: src/main/java/com/civicscience/datapipeline/DataStreamJob.java [hunk 1]
```diff
@@ -122,10 +133,108 @@ public static void main(String[] args) throws Exception {
...
+  }
+
+  /**
+   *
+   * @param profile
+   * @return configured kafka sink for trending urls
+   */
+  private static KafkaSink<UrlEntity> getKafkaSink(Profile profile) {
+    Map<String, String> producerConfigMap = Stream.of(
+            new String[][]{{"transaction.timeout.ms", "600000"}, // 10 mins
```
Comment (dkondratenkocs) [inline src/main/java/com/civicscience/datapipeline/DataStreamJob.java:187]: Do we need to externalize some of the below parameters?

File: src/main/java/com/civicscience/datapipeline/DataStreamJob.java [hunk 2]
```diff
@@ -122,10 +133,108 @@ public static void main(String[] args) throws Exception {
...
           @Override
           public void flatMap(String s, Collector<JotLog> collector) {
-            collector.collect(dataTransform.mapToJotLogObject(s));
+            JotLog log = dataTransform.mapToJotLogObject(s);
+            if (log.getType() != null && log.getSequence() != null && log.getId() != null
+                && log.getSpace() != null) {
+              collector.collect(log);
+            }
           }
         }).map(new JotLogTransformMetricMapper());
```
Comment (dkondratenkocs) [inline src/main/java/com/civicscience/datapipeline/DataStreamJob.java:142]: The best practice is to name operations and put id like
`.name("<put here meningful name to see on graph in job manager>").uid("<you may put any id here to differentiate in metrics and graph>")`

File: src/main/java/com/civicscience/datapipeline/DataStreamJob.java [hunk 3]
```diff
@@ -122,10 +133,108 @@ public static void main(String[] args) throws Exception {
...
+      KafkaSink<UrlEntity> contextUrlSink = getKafkaSink(
+          profile);
+
+      urlStream.keyBy(url -> url)
+          .window(
+              TumblingProcessingTimeWindows.of(
+                  Time.minutes(profile.getTrendingUrlProcessingWindowMinutes())))
+          .aggregate(new UrlLoadsAggregate(),
+              new PopularUrlDetectWindowFunction(profile.getMinimumCountForTrendingUrl()))
+          .sinkTo(contextUrlSink);
```
Comment (dkondratenkocs) [inline src/main/java/com/civicscience/datapipeline/DataStreamJob.java:175]: The best practice is to name operations and put id like
`.name("<put here meaningful name to see on graph in job manager>").uid("<you may put any id here to differentiate in metrics and graph>")`

File: src/main/java/com/civicscience/datapipeline/PopularUrlDetectWindowFunction.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+  /**
+   * @param url          The key for which this window is evaluated.
+   * @param context      The context in which the window is being evaluated.
+   * @param urlLoadsCnts The elements in the window being evaluated.
+   * @param out          A collector for emitting elements.
+   */
+  public void process(String url, Context context, Iterable<Long> urlLoadsCnts,
+      Collector<UrlEntity> out) {
+    Long urlLoadsCnt = urlLoadsCnts.iterator().next();
```
Comment (dkondratenkocs) [inline src/main/java/com/civicscience/datapipeline/PopularUrlDetectWindowFunction.java:23]: if urlLoadsCnts.iterator().next() returns null?

File: src/main/java/com/civicscience/datapipeline/PopularUrlDetectWindowFunction.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    this.minimumCountForTrendingUrl = minimumCountForTrendingUrl;
+  }
+
+  /**
+   * @param url          The key for which this window is evaluated.
+   * @param context      The context in which the window is being evaluated.
+   * @param urlLoadsCnts The elements in the window being evaluated.
+   * @param out          A collector for emitting elements.
+   */
+  public void process(String url, Context context, Iterable<Long> urlLoadsCnts,
```
Comment (dkondratenkocs) [inline src/main/java/com/civicscience/datapipeline/PopularUrlDetectWindowFunction.java:21]: Put javadoc by explaining when this method is called and what is doing

File: src/main/java/com/civicscience/datapipeline/FilePathFilterS3.java [hunk 6]
```diff
@@ -34,21 +42,39 @@ public FilePathFilterS3(Duration ageLimit) {
   @Override
   public boolean filterPath(Path path) {
```
Comment (dkondratenkocs) [inline src/main/java/com/civicscience/datapipeline/FilePathFilterS3.java:43]: How do you think we may refactor this method in future to make it less error prone and clear?

File: src/main/java/com/civicscience/datapipeline/FilePathFilterS3.java [hunk 7]
```diff
@@ -18,11 +18,19 @@ public class FilePathFilterS3 extends FilePathFilter implements Predicate<Path>,
...
   private final Duration ageLimit;
+  private final boolean useRange;
+  private final LocalDate start;
+  private final LocalDate end;
 
-  public FilePathFilterS3(Duration ageLimit) {
+  public FilePathFilterS3(Duration ageLimit, boolean useRange, String start, String end) {
     this.ageLimit = ageLimit;
+    this.useRange = useRange;
+    this.start = LocalDate.parse(start);
```
Comment (dkondratenkocs) [inline src/main/java/com/civicscience/datapipeline/FilePathFilterS3.java:32]: 1. what is the format for start and end?
2. Add checks for null, empty string and so on? StringUtils.isNotBlank - checks for nulls and empty
3. How are we going to handle string that is unparsable

File: src/main/java/com/civicscience/datapipeline/DataStreamJob.java [hunk 8]
```diff
@@ -122,10 +133,108 @@ public static void main(String[] args) throws Exception {
...
+    DataStream<String> urlStream = jotLogDataStream.flatMap(new FlatMapFunction<>() {
+      @Override
+      public void flatMap(JotLog value, Collector<String> out) {
+        if (value.getClean_context_url() != null && !value.getClean_context_url().isEmpty()
+            && !value.getUa_is_bot()
+            && value.getType().equals("resolved")
+            && (value.getPin_mode() != null && (value.getPin_mode().equals("enabled")
+            || value.getPin_mode().equals("required")))
+            && (value.getPin() != null && value.getPin().isEmpty()
+            && NumberUtils.isNumber(value.getTarget_id()))
```
[the remark this answers is not in the document]
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/DataStreamJob.java]: added null check

File: src/main/java/com/civicscience/datapipeline/FilePathFilterS3.java (hunk 7 above)
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/FilePathFilterS3.java:32]: start and end are passed/configurable in flink-params yml, format 2023-01-29. It always has a default value. It is not coming from external system or logs.

File: src/main/java/com/civicscience/datapipeline/DataStreamJob.java [hunk 9]
```diff
@@ -122,10 +133,108 @@ public static void main(String[] args) throws Exception {
...
+                && log.getSpace() != null) {
+              collector.collect(log);
+            }
           }
         }).map(new JotLogTransformMetricMapper());
 
+    DataStream<String> urlStream = jotLogDataStream.flatMap(new FlatMapFunction<>() {
+      @Override
+      public void flatMap(JotLog value, Collector<String> out) {
+        if (value.getClean_context_url() != null && !value.getClean_context_url().isEmpty()
```
[the remark this answers is not in the document]
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/DataStreamJob.java]: I added comments to explain

File: src/main/java/com/civicscience/datapipeline/UrlLoadsAggregate.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
+package com.civicscience.datapipeline;
+
+import org.apache.flink.api.common.functions.AggregateFunction;
+
+public class UrlLoadsAggregate implements AggregateFunction<String, Long, Long> {
```
[the remark this answers is not in the document]
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/UrlLoadsAggregate.java:9]: done

File: src/main/java/com/civicscience/datapipeline/PopularUrlDetectWindowFunction.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
+package com.civicscience.datapipeline;
+
+import com.civicscience.entity.UrlEntity;
+import org.apache.flink.streaming.api.functions.windowing.ProcessWindowFunction;
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
+import org.apache.flink.util.Collector;
+
+public class PopularUrlDetectWindowFunction extends
```
[the remark this answers is not in the document]
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/PopularUrlDetectWindowFunction.java:8]: done

File: src/main/java/com/civicscience/datapipeline/FilePathFilterS3.java (hunk 6 above)
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/FilePathFilterS3.java:43]: what errors do you see?

File: src/main/java/com/civicscience/datapipeline/DataStreamJob.java (hunk 1 above)
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/DataStreamJob.java:188]: yeah we can

File: src/main/java/com/civicscience/datapipeline/DataStreamJob.java (hunk 3 above)
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/DataStreamJob.java:176]: adding

File: src/main/java/com/civicscience/datapipeline/DataStreamJob.java (hunk 1 above)
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/DataStreamJob.java]: done

File: src/main/java/com/civicscience/datapipeline/DataStreamJob.java [hunk 12]
```diff
@@ -122,10 +133,108 @@ public static void main(String[] args) throws Exception {
...
+
+    if (profile.isSinkToS3Enabled()) {
+      s3Sink(profile, jotLogDataStream);
+    }
+
+    if (profile.isSinkToKafkaEnabled()) {
+      KafkaSink<UrlEntity> contextUrlSink = getKafkaSink(
+          profile);
+
+      urlStream.keyBy(url -> url)
```
[the remark this answers is not in the document]
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/DataStreamJob.java:173]: phase-2, I will create a Jira task

[the remark this answers is not in the document]
Comment (shanmukhipriya-ponnada) [reply src/main/java/com/civicscience/datapipeline/DataStreamJob.java:173]: https://civicscience.atlassian.net/browse/DATA-409

File: src/main/resources/default-flink-params.yml [hunk 13]
```diff
@@ -25,12 +53,39 @@ prod:
...
+    sink_to_s3_enabled: false
+    sink_to_kafka_enabled: true
+
     filesink_parallelism: 5
     filesink_bucket_name: "civicscience-dwf-prod-sink"
     filesink_input_path: "/prod-sink"
     filesink_part_size: 50
 
+    confluent_api_key: "DCVXAUKN5ROAKOHP"
+    confluent_api_secret: "VF8dKAJLqBK+ZqInybUieG+s9sUpnKmyUyx5k+uGlpw7FO8clmM/eLkPbcV3vUJY"
```
Comment (dkondratenkocs) [inline src/main/resources/default-flink-params.yml:72]: Candidate fro secret store

**Comment by ychen-civicscience** [comment]
Looks good to me.

File: src/main/resources/default-flink-params.yml (hunk 13 above)
Comment (shanmukhipriya-ponnada) [reply src/main/resources/default-flink-params.yml:72]: We will implement it in next phase

**Pull request merged**